src/net_processing.cpp

```diff
@@ -1538,6 +1538,13 @@ void static ProcessGetData(CNode* pfrom, const CChainParams& chainparams, CConnm
             if (mi != mapRelay.end()) {
                 connman->PushMessage(pfrom, msgMaker.Make(nSendFlags, NetMsgType::TX, *mi->second));
                 push = true;
+            } else if (!g_relay_txes && !pfrom->HasPermission(PF_RELAY)) {
+                // If a blocks-only peer requests a tx and it isn't in our mapRelay, then disconnect them.
+                // We allow blocks-only peers to request txs in our mapRelay so we can relay txs submitted
+                // locally when in blocks-only mode.
+                LogPrintf("Blocks-only peer %d sent us tx GETDATA in violation of protocol\n", pfrom->GetId());
```

Comment:
I believe this is incorrect -- it should not be a protocol violation if we announce a transaction to a peer and it happens to wait more than 15 minutes to respond with a GETDATA.  This already happens for reasons that are benign (due to us not immediately requesting a transaction from all peers who announce it).  

For example: one way this can happen is if we announce a transaction to a peer, but it gets the announcement as well from other peers first and queues up a request to go to us far in the future, and in the meantime a block is found that includes the transaction as well as spends of its outputs, and then at some point when it's time for the peer to consider requesting the tx from us it has at that point forgotten about the tx completely and it'll re-request.  (This type of behavior is present in all the version of Bitcoin Core I can remember.)

I think to do this right we would need to either track the tx's we've announced to a peer, or consider changing blocks-only behavior so that we never announce anything at all (which makes more logical sense to me, but I think others disagree, so probably a non-starter).